packages/mermaid/src/diagrams/xychart/xychartDb.ts

```diff
@@ -0,0 +1,302 @@
+function setYAxisRangeFromPlotData(data: number[]) {
+  const minValue = Math.min(...data);
+  const maxValue = Math.max(...data);
+  const prevMinValue = isLinearAxisData(xyChartData.yAxis) ? xyChartData.yAxis.min : Infinity;
+  const prevMaxValue = isLinearAxisData(xyChartData.yAxis) ? xyChartData.yAxis.max : -Infinity;
+  xyChartData.yAxis = {
+    type: 'linear',
+    title: xyChartData.yAxis.title,
+    min: Math.min(prevMinValue, minValue),
+    max: Math.max(prevMaxValue, maxValue),
+  };
+}
+
+function transformDataWithOutCategory(data: number[]): SimplePlotDataType {
+  let retData: SimplePlotDataType = [];
+  if (data.length === 0) {
+    return retData;
+  }
+  if (!hasSetXAxis) {
+    const prevMinValue = isLinearAxisData(xyChartData.xAxis) ? xyChartData.xAxis.min : Infinity;
+    const prevMaxValue = isLinearAxisData(xyChartData.xAxis) ? xyChartData.xAxis.max : -Infinity;
+    setXAxisRangeData(Math.min(prevMinValue, 1), Math.max(prevMaxValue, data.length));
+  }
+  if (!hasSetYAxis) {
+    setYAxisRangeFromPlotData(data);
+  }
+
+  if (isBandAxisData(xyChartData.xAxis)) {
+    retData = xyChartData.xAxis.categories.map((c, i) => [c, data[i]]);
+  }
+
+  if (isLinearAxisData(xyChartData.xAxis)) {
+    const min = xyChartData.xAxis.min;
+    const max = xyChartData.xAxis.max;
+    const step = (max - min + 1) / data.length;
+    const categories: string[] = [];
+    for (let i = min; i <= max; i += step) {
+      categories.push(`${i}`);
+    }
+    retData = categories.map((c, i) => [c, data[i]]);
+  }
+
+  return retData;
+}
+
+let plotIndex = 0;
+
+function getPlotColorFromPalette(plotIndex: number): string {
+  return plotColorPalette[plotIndex === 0 ? 0 : plotIndex % (plotColorPalette.length - 1)];
+}
+
+function setLineData(title: NormalTextType, data: number[]) {
+  const plotData = transformDataWithOutCategory(data);
+  xyChartData.plots.push({
+    type: 'line',
+    strokeFill: getPlotColorFromPalette(plotIndex),
+    strokeWidth: 2,
+    data: plotData,
+  });
+  plotIndex++;
+}
+
+function setBarData(title: NormalTextType, data: number[]) {
+  const plotData = transformDataWithOutCategory(data);
+  xyChartData.plots.push({
+    type: 'bar',
+    fill: getPlotColorFromPalette(plotIndex),
+    data: plotData,
+  });
+  plotIndex++;
+}
+
+function getDrawableElem(): DrawableElem[] {
+  if (xyChartData.plots.length === 0) {
+    throw Error('No Plot to render, please provide a plot with some data');
+  }
+  xyChartData.title = getDiagramTitle();
+  return XYChartBuilder.build(xyChartConfig, xyChartData, xyChartThemeConfig);
+}
+
+function setHeight(height: number) {
+  xyChartConfig.height = height;
+}
+
+function setWidth(width: number) {
+  xyChartConfig.width = width;
+}
+
+const clear = function () {
+  commonClear();
+  plotIndex = 0;
+  xyChartConfig = getChartDefaultConfig();
+  xyChartData = getChartDefalutData();
+  xyChartThemeConfig = getChartDefaultThemeConfig();
```

Comment:
This means the data being set could be modified.
Use structuredClone to make sure a copy is taken.

Reply:
Spread operator will only do a shallow clone. We need to use `structuredClone` (for just data), or something like lodash (if there are functions as properties) to do a proper deep clone. `structuredClone` is a built in function, so there should be no major performance impact. 

[Example](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAhjAvDA3jAUDGAPAXDAcmwIBpMYBPfFcrASzDAFMAnfA4A8gX3V-VCRYEAA4smcACZJUAOnlxeg6DGAAbcExnQWAV2BRd4yQGENzABRwAlOgHmms7DI4ALOGADmTSQQDc9pqyFLIMzCwuwO5ePlzKIGqOGp4WKHAkqg4ZouJS3LY5EpJOkdHekgCCnnAMXIVSwaGMrKUe5VU1YHHgEAlJIClp2WJFGeqa+ehAA)

```ts
const a = { 
  x: 'x',
  y: {
    inner: 'c'
  }
}

const spread = {...a}
const clone = structuredClone(a)

clone.x = 'changed';
clone.y.inner = 'changed'
console.log({a, clone, spread})
spread.x = 'changedAgain'
spread.y.inner = 'changedAgain'
console.log({a, spread, clone})
```

```
[LOG]: {
  "a": {
    "x": "x",
    "y": {
      "inner": "c"
    }
  },
  "clone": {
    "x": "changed",
    "y": {
      "inner": "changed"
    }
  },
  "spread": {
    "x": "x",
    "y": {
      "inner": "c"
    }
  }
} 
[LOG]: {
  "a": {
    "x": "x",
    "y": {
      "inner": "changedAgain"
    }
  },
  "spread": {
    "x": "changedAgain",
    "y": {
      "inner": "changedAgain"
    }
  },
  "clone": {
    "x": "changed",
    "y": {
      "inner": "changed"
    }
  }
} 
```

Reply:
I got your point, For the data whenever we are calling `getChartDefalutData();` every time new data is generated so no need for a clone.

for `getChartDefaultThemeConfig();` the structure is 1 level deep so spread operator is fine for this.
for `getChartDefaultConfig();` the structure is 2 level deep and I am doing spread on the inner object also like
```javascript
const spread = {...a, y: {...{inner: "c"}}}
```

Real implementation
```javascript
function getChartDefaultConfig(): XYChartConfig {
  return {
    ...(defaultConfig.xyChart as XYChartConfig),
    ...config.xyChart,
    yAxis: {
      ...(defaultConfig.xyChart as XYChartConfig).yAxis,
      ...config.xyChart?.yAxis,
    },
    xAxis: {
      ...(defaultConfig.xyChart as XYChartConfig).xAxis,
      ...config.xyChart?.xAxis,
    },
  };
}
```

Thats why I think adding `structuredClone` will be redundant, Please let me know if I am missing anything.

Reply:
Doing this clone manually is unnecessary and not maintainable. If someone later adds a new config and doesn't clone it properly, things could break. 

`structuredClone` should be the first choice, `cleanAndMerge` the second. 

<img width="538" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/31c13107-b62d-45bd-a3e2-675158fc6058">

Reply:
Since Sid had already approved this PR, I am closing this as resolved